### PR TITLE
[rush] Add support for "global" custom commands

### PR DESCRIPTION
**Pull request by pgonzal**
This PR allows global commands to be defined in command-line.json.  These commands are executed once for the whole repository.

### CODE REVIEW

**Comment by pgonzal** [comment]
@ThomasMichon FYI -- this will support scenarios such as "create-pullrequest" in odsp-common

File: apps/rush-lib/src/cli/scriptActions/GlobalScriptAction.ts [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+  ) {
+    super(options);
+    this._scriptPath = options.scriptPath;
+  }
+
+  public run(): Promise<void> {
+    if (!fsx.existsSync(this.rushConfiguration.rushLinkJsonFilename)) {
+      throw new Error(`File not found: ${this.rushConfiguration.rushLinkJsonFilename}` +
+        `${os.EOL}Did you run "rush link"?`);
+    }
```
Comment (nickpape-msft) [inline apps/rush-lib/src/cli/scriptActions/GlobalScriptAction.ts]: this should be optional

File: apps/rush-lib/src/cli/scriptActions/BaseScriptAction.ts [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+export interface IBaseScriptActionOptions extends IBaseRushActionOptions {
+  commandLineConfiguration: CommandLineConfiguration;
+}
+
+/**
+ * Base class for command-line actions that are implemented using user-defined scripts.
+ *
+ * @remarks
+ * Compared to the normal built-in actions, these actions are special because (1) they
+ * can be discovered dynamically via from common/config/command-line.json, and (2)
```
Comment (nickpape-msft) [inline apps/rush-lib/src/cli/scriptActions/BaseScriptAction.ts]: >via from [](start = 33, length = 8)

?

File: apps/rush-lib/src/cli/scriptActions/GlobalScriptAction.ts [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+        }
+      );
+
+      if (exitCode > 0) {
+        console.log(os.EOL + colors.red(`The script failed with exit code ${exitCode}`));
+      }
+
+      process.exitCode = exitCode;
+
+      throw new AlreadyReportedError();
```
Comment (nickpape-msft) [inline apps/rush-lib/src/cli/scriptActions/GlobalScriptAction.ts]: >throw new AlreadyReportedError(); [](start = 6, length = 33)

this should be inside the exitCode > 0 block

File: apps/rush-lib/src/cli/scriptActions/GlobalScriptAction.ts (hunk 1 above)
Comment (pgonzal) [reply apps/rush-lib/src/cli/scriptActions/GlobalScriptAction.ts]: After chatting we decided to remove this check entirely.

---
In reply to: [194200168](https://github.com/Microsoft/web-build-tools/pull/684#discussion_r194200168) [](ancestors = 194200168)

**Pull request merged**